### ISSUE TITLE
Return type of `__getattr__` is not respected for attributes set with custom `__setattr__`

### ISSUE THREAD

**Issue by evtn** [microsoft/pyright#5121]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When setting attribute and the class has an overloaded `__setattr__`, pyright still assumes that attribute is set and ignores `__getattr__`.

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**
I understand that it is a fairly complex issue since there's no way to check with types that the attribute in question will be actually set, and this could lead to wrong typing of existing attributes, but I'd like to know my options here, apart from `# type: ignore`

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.

```python
from typing import Any

class A:
    def __setattr__(self, __name: str, __value: Any) -> None:
        pass

    def __getattr__(self, __attr: str) -> int:
        return 10

a = A()
a.test = "anything"

x = a.test + 120 # error here, a.test is supposedly `Literal['anything']`

```

**Additional context**
My use-case is [this library](https://github.com/courage-tci/lark_dynamic#basic-usage), where setting an attribute puts it into an internal storage, and attribute retrieval returns a proxy class that acts differently (so not the same type). 

Despite this relation being typed, Pyright still complains (for reasons I've outlined above).

**Comment by erictraut** [comment]
This issue reminds me of a discussion we had earlier about properties (and more generally descriptor objects) that are "asymmetric" — i.e. their getters and setters disagree on type. After much discussion in the typing forums, we (the typing community) collectively decided that type checkers should not narrow the type on assignment in cases where an asymmetric descriptor is the target. It is safe (and generally desirable) to narrow in cases where a symmetric descriptor is used.

Pyright currently has no similar logic for detecting asymmetries between `__getattr__` and `__setattr__`. It appears that mypy does detect this and avoid narrowing in this case. I think that approach makes sense and is consistent with the previous resolution for asymmetric descriptors.

**Comment by erictraut** [comment]
This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.310, which I just published. It will also be included in a future release of pylance.
